pandas/core/generic.py

```diff
@@ -489,16 +489,51 @@ def _get_index_resolvers(self):
             d.update(self._get_axis_resolvers(axis_name))
         return d
 
-    def _get_space_character_free_column_resolvers(self):
-        """Return the space character free column resolvers of a dataframe.
+    def _get_special_character_free_column_resolvers(self, expr: str) -> dict:
+        """Return the special character free column resolvers of a dataframe.
 
-        Column names with spaces are 'cleaned up' so that they can be referred
-        to by backtick quoting.
+        Column names with special characters are 'cleaned up' so that they can
+        be referred to by backtick quoting.
         Used in :meth:`DataFrame.eval`.
-        """
-        from pandas.core.computation.common import _remove_spaces_column_name
 
-        return {_remove_spaces_column_name(k): v for k, v in self.items()}
+        Raises
+        ------
+        ValueError if the 'expr' uses names that cannot be unambigiously be
+        resolved from the column names.
+        """
+        from pandas.core.computation.expr import (
+            _BACKTICK_QUOTED_STRING,
+            _create_valid_python_identifier,
+            _tokenize_string,
+            _tokenize_and_clean_column_names,
+        )
+
+        # Find out if the used names in the expr, cannot be unambigiously be
+        # retrieved from the column names after they have been 'cleaned',
+        # and raise a ValueError if that is a problem.
+        cleaned_names = (_tokenize_and_clean_column_names(name) for name in self.keys())
+        # Only check the backtick variables used in the expression.
+        backtick_variables_expr = {
+            _create_valid_python_identifier(tokval)
+            for toknum, tokval in _tokenize_string(expr)
+            if toknum == _BACKTICK_QUOTED_STRING
+        }
+        cleaned_names_expr = (
```

Comment:
Can you use a set difference instead of another loop?

Reply:
I would rather lose that information and keep the implementation and corresponding error message much simpler

Reply:
I need that information to know if I need to raise the error message. I need to know if there is name with count > 1. If that is the case, I need to raise the error message. If not, I don't have to. 

However, I have changed the iterator comprehension to a map / filter combination. Maybe that is a bit more readable than the `name in name for ... if name in ...`.

Reply:
Take a look at the new commits.